### DISCUSSION TITLE
Subscribing to assets

### DISCUSSION THREAD

**Discussion by robot-ranger** [mtconnect/cppagent#478] [Q&A]
Question about cppagent capability: can we open a stream on /assets endpoint? Or is there some other "subscribe"-like mechanism for assets? Such as opening a stream on /sample, but for assets?

Were implementing Task assets and looking to avoid polling if possible.

**Comment by wsobel** [comment]
There are two data items for each device AssetChanged and AssetRemoved. Changed handles all adds and updates, removed handles removals where the asset is marked for removal. 

If you listen to those two data items it should do what you want. Once you get the asset id from the event, you can query for the new asset.

**Comment by robot-ranger** [comment]
That doesnt seem to scale. What happens when there are a sufficiently large number of devices? All devices have to be listening to all other devices?

**Comment by robot-ranger** [comment]
to be clear, i am asking this int the context of tasks assets

**Comment by wsobel** [comment]
This works the same for all asset types. Each device will need to listen to other devices in their domain. It will scale since the amount of data will be very small if you only subscribe to a few event types. The issue in a distributed system is you need to have peer to peer connections. This is where we can consider discovery as a necessary part of this solution. 

We should discuss discovery as well.

**Comment by robot-ranger** [comment]
Can you walk me through how a `MATERIAL_UNLOAD` collaborator would discover a task it can do?

Lets take a potential robotic machine tender collaborator looking for a task to do for example. 
we were expecting to "watch" the `/assets` end point and filter them by criteria (asset that is a task w/ type `MATERIAL_UNLOAD` and state `INACTIVE` or `PREPARING`)

Instead:
1) "subscribe" to `/sample?path=//DataItem[@type=%22ASSET_CHANGED%22]`
2) when anyone in the fleet represents `ASSET_CHANGED`, we capture that asset id, then query `/assets/my_asset_id`
3) determine if we care (ie - is it a task? are we capable? is it not already committed?)

do i have that right?
